### PR TITLE
Sage Detection and ID Workflow Hardening

### PR DESCRIPTION
**Pull request by bluemellophone**
This PR incorporates overall improvements for the detection and ID workflows as it communicates with Sage.  This PR is a continuation of https://github.com/WildMeOrg/houston/pull/631

Known Regressions:
* The following tests were disabled for using too many resources, will be enabled again in a future PR:
  * integration_tests/test_sage_detection.py::test_create_asset_group_detection
  * integration_tests/test_sage_basic_identification.py::test_create_asset_group_identification
  * integration_tests/test_sage_basic_identification.py::test_identification_international

---

* Rename ACM to Sage
  * app/extensions/acm -> app/extensions/sage
  * Updated the Sage extension to set `USE_JSON_HEADERS=False` and add many more endpoints:
``` python 
    ENDPOINTS = {
         # No user.session, Sage doesn't support logins
         'annotation': {
             'list': '//annot/json/',
             'data': '//annot/name/uuid/json/?annot_uuid_list=[{"__UUID__": "%s"}]',
             'exists': '//annot/rowid/uuid/json/?annot_uuid_list=[{"__UUID__":"%s"}]',
             'create': '//annot/json/',
             'delete': '//annot/json/',
         },
         'asset': {
             'list': '//image/json/',
             'create': '//image/json/',
             'exists': '//image/rowid/uuid/json/?image_uuid_list=[{"__UUID__":"%s"}]',
             'upload': '//upload/image/json/',
             'delete': '//image/json/',
         },
         'version': {
             'dict': '//version/',
         },
         'passthrough': {
             'data': '',
         },
         'engine': {
             'list': '//engine/job/status/',
             'detect': '//engine/detect/cnn/',
             'identification': '//engine/query/graph/',
             'result': '//engine/job/result/?jobid=%s',
             'status': '//engine/process/status/',
         },
     }
```

* Add `sage-sync` Docker service, which is used for sending assets and annotations (@mmulich note that you'll need to specify this in production)
* Added `sage.SageModel` to give specific model classes access to Sage-specific functionality (i.e., syncing)
  * Added `sync_with_sage` to `Asset` to sync an asset and ensure it exists within Sage
  * Added `sync_with_sage` to `Annotation` to sync an annotation and ensure it exists within Sage (also its Asset)
  * Removed `ensure_sage` on `Annotation` as it was now duplicate functionality
* Updated the syncing of Assets and Annotations to a push design instead of pull.  This allows the processing flow to remain within Houston, which has control of errors and can respond accordingly when things fail to send to Sage.  Also, pushing content into Sage allows us to quickly ask Sage if it already exists, preempting the need to have Sage re-download images to make the same determination.
* Added database migration `afedda6596b9`

---
 
* Refactor all Celery tasks to prevent using `.delay` during testing and avoid calling the task function as well.  If working in the foreground, all processing and control should remain in the foreground process.
* Added ahead tracking to Progress for Celery tasks and Sage tasks in their respective queues.
* Added parents to Progress to allow steps to be implemented
* Added `progress_detection` and `progress_identification` to `GitStore`, `AssetGroupSighting`
* Added `progress_identification` to `Annotation`, 'Sighting'
* Refactored Sightings to send an `Annotation` object to `send_identification` instead of string GUIDs.
* Added pre-send checks to Sighting send ID call to ensure that the requests is semantically correct.
* Added Sage-specific tasks to get the current status, ensure Assets and Annotations have been synced, and force syncing
* Added Celery task to keep Sage in sync, pruned, and check job statuses

---

* Since we can pre-send content to Sage, we can run tests with actual detection and ID responses with Sage.  Removed all of the Mock code for Sage in testing and removed all simulated Sage requests and responses
* Added `tests.utils.wait_for_progress()` to allow tests to wait for Sage processes and automatically pull their responses from the server when completed
* Fix all test warnings

---

* Fixes the 413 communication error with Sage
* Add JSON file linting with `pre-commit` (validating, formatting, and cleaning up all `ia-configs` files)
* Standardize all usage of `datetime` to use a global import without namespacing for `datetime.datetime` and `datetime.timedelta`
* Removed `/api/v1/dev` API namespace, not used
* Removed task `invoke app.dev.embed`, not used
* Moved EDM and Sage tasks to root level, i.e. `invoke edm.*` and `invoke sage.*`
* Updated all variables and parameters to use `foreground` over `background` when allowing a condition for an optional Celery task vs. foreground compute.

### CODE REVIEW

File: app/extensions/git_store/__init__.py
Comment (bluemellophone) [reply app/extensions/git_store/__init__.py:615]: This is a good idea, but for now I want these to be separate functions.  We can consolidate in a future PR if we want to clean it up.

File: app/modules/annotations/models.py
Comment (bluemellophone) [inline app/modules/annotations/models.py:362]: Is this filter still needed?  My guess is no because if an Annotation or Asset doesn't have a content_guid, we will create it as soon as we need to send it to Sage

File: app/modules/asset_groups/models.py
Comment (bluemellophone) [inline app/modules/asset_groups/models.py:490]: This is now using a better default value, making this parameter not needed and results in the same behavior

File: app/modules/asset_groups/models.py
Comment (bluemellophone) [inline app/modules/asset_groups/models.py:795]: This is not needed.  It is sent to Sage and is ignored.  The endpoint is encoded within the URL, not the body of the request

File: app/modules/asset_groups/models.py
Comment (bluemellophone) [inline app/modules/asset_groups/models.py:847]: If `preload=True`, we preemptively send the Assets to Sage by uploading its file bytes to Sage.  This returns a UUID and we store that on the Asset as its content_guid.  If the content_guid is specified on an asset, we ensure that it exists by checking with Sage.  If it doesn't, `asset.sync_with_sage(ensure=True)` will reset and force push the Asset to Sage.

File: app/modules/asset_groups/models.py
Comment (bluemellophone) [inline app/modules/asset_groups/models.py:859]: If `preload=False`, we send the Houston asset URLs to Sage, which will download the assets from Houston.  This is not preferred for two reasons: 1) We don't want to send the file bytes to Sage if it already has the asset (which Sage will have to download the file to determine if it already has the asset) and 2) The asset may have been deleted or access otherwise disabled between the time Sage received the request and actually attempts to download the files.  Whevener possible, we should front-load the sending of Assets and Annotations to Sage because Houston maintains control of the processing flow and can respond to errors in the foreground process

File: app/modules/asset_groups/models.py
Comment (bluemellophone) [inline app/modules/asset_groups/models.py:1046]: This has been moved to the `init_progress_detection` call.

File: app/modules/assets/models.py
Comment (bluemellophone) [inline app/modules/assets/models.py:315]: The `is_detection()` check was preventing Sage from downloading the asset when it needed it.  Opening this up is a security consideration, but is limited to internal users.

File: app/modules/assets/models.py
Comment (bluemellophone) [inline app/modules/assets/models.py:482]: Fixes a PIL deprecation warning

File: app/modules/sightings/models.py
Comment (bluemellophone) [inline app/modules/sightings/models.py:1234]: Store the Sage job UUID for the ID task

File: app/modules/sightings/schemas.py
Comment (bluemellophone) [inline app/modules/sightings/schemas.py:190]: Standardize the output with the detection API

File: ia-configs/IA.GLOBALS.json
Comment (bluemellophone) [inline ia-configs/IA.GLOBALS.json:8]: I have updated the Sage image to combine this into a single API, `/api/engine/detect/cnn/`.  Houston can specify `algo` as a parameter to use either `lightnet` or `yolo`.

File: ia-configs/IA.GLOBALS.json
Comment (bluemellophone) [inline ia-configs/IA.GLOBALS.json:6]: This needs to be reviewed, is it being used?  @drewblount

File: ia-configs/IA.whaleshark.json
Comment (bluemellophone) [inline ia-configs/IA.whaleshark.json:7]: This new attribute `algo` is used to send the detection job to the older YOLO detector

File: tests/modules/assets/resources/test_assets.py
Comment (bluemellophone) [inline tests/modules/assets/resources/test_assets.py:174]: Yes, we do want to allow this for internal users

File: tests/modules/sightings/test_models.py
Comment (bluemellophone) [inline tests/modules/sightings/test_models.py:114]: This test is covered by a different test that actually runs the jobs and checks their results.  There is no need to check the simulated accuracy of only the job schema

**Review by bluemellophone — COMMENTED** [review]
LGTM, self-approved

File: app/modules/annotations/models.py
Comment (naknomum) [reply app/modules/annotations/models.py:362]: i am going to agree with this.  not coming up with a reason we would want Annotations to get sent to Sage and assigned a `content_guid`.   (i can think of some very edge-y cases for Assets; but that is irrelevant here.)

File: app/modules/asset_groups/models.py
Comment (naknomum) [inline app/modules/asset_groups/models.py:109]: thinking this should stay?

Comment (bluemellophone) [reply app/modules/asset_groups/models.py:109]: Agreed, I've added it back in where `ags.setup()` is called after creation

File: integration_tests/test_sage_basic_identification.py
Comment (bluemellophone) [inline integration_tests/test_sage_basic_identification.py:79]: I'm debugging high resource usage during integration tests

File: app/modules/annotations/models.py
Comment (bluemellophone) [reply app/modules/annotations/models.py:360]: Agreed, removing the commented-out filter

File: app/modules/annotations/models.py
Comment (drewblount) [inline app/modules/annotations/models.py:183]: If we're ever doing a reformatting PR later, these two ifs could be a single if ___ and ___. Pretty minor code style thing so not blocking this PR.

File: app/modules/assets/models.py
Comment (drewblount) [inline app/modules/assets/models.py:285]: second annoying style comment from me on the double ifs

**Review by naknomum — APPROVED** [review]
what just happened!?  haha  looks good, but i have some usage/learning to do for sure.

File: ia-configs/IA.amphibian-reptile.json
Comment (drewblount) [inline ia-configs/IA.amphibian-reptile.json:50]: Since these notes are meant to be read by a programmer opening the config file I prefer them at the top. Not a big deal.

**Review by drewblount — APPROVED** [review]
This is huge! But LGTM if all the tests are good. I like the asset/annot syncing and love that we aren't simulating sage tests anymore.

Also a good example of "if you do enough good work you can force through any style changes you've been wanting" with the datetime imports and renaming sage, hah. I like those QoL improvements.

My comments on the ia_config.json file formatting could be addressed but I'm OK without them: It'd be nice to have _note and _readme comments at the top of our config files. In slack convo with JP discussed formatting these just the once then excluding them from pre-commit.

File: app/modules/assets/models.py
Comment (bluemellophone) [reply app/modules/assets/models.py:285]: Consolidated

File: ia-configs/IA.zebra.json
Comment (bluemellophone) [reply ia-configs/IA.zebra.json:26]: Fixed

File: app/modules/annotations/models.py
Comment (bluemellophone) [reply app/modules/annotations/models.py:183]: Consolidated

File: ia-configs/IA.amphibian-reptile.json
Comment (bluemellophone) [reply ia-configs/IA.amphibian-reptile.json:50]: The pre-commit options for [`pretty-format-json`](https://gdevops.gitlab.io/tuto_git/tools/pre-commit/repos_hooks/repo_pre_commit_hooks/pretty-format-json/pretty-format-json.html) has an argument for `--top-keys`.   This argument allows me to sort the list of keys but specify a list of keys to keep near the top.  I've added `_README` and `_note`.

**Pull request merged**
